### PR TITLE
[event] support WithField option in RAY_EVENT api

### LINKED ISSUE #16252: [Dashboard][event] Custom field API support.
- We already has "custom_fields" attribute in event log. But no API to set fields.

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?

like this:
```
  RAY_EVENT(INFO, "label 1")
          .WithField("string", "test string")
          .WithField("int", 123)
          .WithField("double", 0.123)
          .WithField("bool", true)
      << "send message 1";
```

## Related issue number

Closes #16252

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/util/event.h [hunk 1]
```diff
@@ -181,6 +191,7 @@ class RayEvent {
  private:
   rpc::Event_Severity severity_;
   std::string label_;
+  std::unordered_map<std::string, std::string> custom_fields_;
```
Comment (scv119) [inline src/ray/util/event.h:194]: will we add support for integral/double types? it will be useful for recording and analyze

**Review by scv119 — DISMISSED** [review]
thanks for contributing! Though this seems simple enough, can you add unit test to cover this change?

File: src/ray/util/event.h [hunk 2]
```diff
@@ -164,6 +164,16 @@ class RayEvent {
     return *this;
   }
 
+  RayEvent &WithField(const std::string &key, const std::string &value) {
+    custom_fields_[key] = value;
+    return *this;
+  }
+  RayEvent &WithFields(
+      const std::unordered_map<std::string, std::string> &custom_fields) {
+    custom_fields_ = custom_fields;
```
Comment (scv119) [inline src/ray/util/event.h:173]: hmm, should we merge the values into custom_fields other than replacement?

Comment (rkooo567) [reply src/ray/util/event.h:173]: Yeah I think this is a better UX. Ideally, we should set the custom fields once in the initialization (per component) and add withFields for each event.

**Review by rkooo567 — COMMENTED** [review]
So, TODO seems like 

- Support int / double types
- Merging instead of replacement for custom fields
- Unit test

File: src/ray/util/event.cc [hunk 3]
```diff
@@ -57,10 +57,22 @@ LogEventReporter::~LogEventReporter() { Flush(); }
 
 void LogEventReporter::Flush() { log_sink_->flush(); }
 
-std::string LogEventReporter::EventToString(const rpc::Event &event) {
-  std::stringstream result;
+std::string LogEventReporter::replaceLineFeed(std::string message) {
+  std::stringstream ss;
+  // If the message has \n or \r, we will replace with \\n
```
Comment (rkooo567) [inline src/ray/util/event.cc:62]: Why do we need this? Is it for json serialization?

File: src/ray/util/event.h [hunk 4]
```diff
@@ -164,6 +172,12 @@ class RayEvent {
     return *this;
   }
 
+  template <typename T>
+  RayEvent &WithField(const std::string &key, const T &value) {
+    custom_fields_[key] = value;
```
Comment (rkooo567) [inline src/ray/util/event.h:177]: What's going to happen if we try calling the unsupported types for json?

File: src/ray/util/event.cc [hunk 5]
```diff
@@ -57,10 +57,22 @@ LogEventReporter::~LogEventReporter() { Flush(); }
...
+      ss << "\\n";
+    } else {
+      ss << message[i];
+    }
+  }
+  return ss.str();
+}
 
-  boost::property_tree::ptree pt;
+std::string LogEventReporter::EventToString(const rpc::Event &event,
```
Comment (rkooo567) [inline src/ray/util/event.cc:73]: Btw, do we need this? Why don't we just store the json serialized string instead of copying to the protobuf first and re-copy to the json string?

**Review by rkooo567 — APPROVED** [review]
I think this LGTM. Have a couple comments

File: src/ray/util/event.cc (hunk 5 above)
Comment (SongGuyang) [reply src/ray/util/event.cc:73]: Yep, this is a problem of history. In the first version of event framework, the event could be transform directly by RPC. But now, we just write an event log. We should remove the protobuf in future. I have added an TODO https://github.com/ray-project/ray/pull/17476/files#diff-01dea7b9807ba283ccf8ffd3131551fd30e0b6d74402f8b07253e256a5b28469R96.

File: src/ray/util/event.cc (hunk 3 above)
Comment (SongGuyang) [reply src/ray/util/event.cc:62]: Make sure the serialized json is one line in the event log. Agent will collect the event log line by line.

File: src/ray/util/event.h (hunk 4 above)
Comment (SongGuyang) [reply src/ray/util/event.h:177]: The compiling will fail. Like this 
```
external/nlohmann_json/include/nlohmann/json.hpp:2174:17: note:   no known conversion for argument 1 from ‘const ray::test_cls’ to ‘nlohmann::basic_json<>’
```

File: src/ray/util/event.cc (hunk 3 above)
Comment (rkooo567) [reply src/ray/util/event.cc:62]: Does this mean the format has tight coupling with the agent implementation here? 

I am asking it because we might want to just scrape the log file and store to some backend directly instead of reporting it to the dahsboard (not confirmed, but it is one of options). In this case, wouldn't this break the json format?

Comment (SongGuyang) [reply src/ray/util/event.cc:62]: Yep, a little hackly here. In our case, when we catch a backtrace to event, the event log written multiple lines. It is hard to parse. We just recover the message content in dashboard. Maybe we can optimize this in future? It is easy to remove this logic.
